keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,969 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Introduce new set of APIs to express an administrator's intent in securing
+their K8s cluster. This doc proposes two new set of resources,
+ClusterNetworkPolicy API and the DefaultNetworkPolicy API to complement the
+developer focused NetworkPolicy API in Kubernetes.
+
+## Motivation
+
+Kubernetes provides NetworkPolicy resources to control traffic within a
+cluster. NetworkPolicies focus on expressing a developers intent to secure
+their applications. Thus, in order to satisfy the needs of a security admin,
+we propose to introduce new set of APIs that capture the administrators intent.
+
+### Goals
+
+The goals for this KEP are to satisfy two key user stories:
+1. As a cluster administrator, I want to enforce irrevocable guardrails that
+   all workloads must adhere to in order to guarantee the safety of my clusters.
+2. As a cluster administrator, I want to deploy a default set of policies to all
+   workloads that may be overridden by the developers if needed.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Backwards compatibility with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of. Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+
+Additionally, this proposal is squarely focused on solving the needs of the
+Cluster Administrator. It is not intended to solve:
+1. Logging / error reporting for Network Policy
+2. Kubernetes Node policies
+3. New policy selectors (services, service accounts, etc.)
+4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
+   impact on workloads.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following two new resources
+under `networking.k8s.io` API group:
+- ClusterNetworkPolicy
+- DefaultNetworkPolicy
```

Comment:
I think we can all agree that both resources speak to the same persona, and that's why we put them in the same KEP. We also started out with a single ClusterNetworkPolicy resource, and defined a `defaultAllow` action. But there was one problem: the rule semantics are different between `defaultAllow` and `allow`/`deny`.  `defaultAllow` won't make sense if there are no implicit isolation enforced, since all Pods should be able to communicate with each other already, without policies. `allow`/`deny` however, shouldn't create any implicit rules and can be read as-is. 

This caused a lot of confusion when we presented it in sig-net bi-weekly (https://www.youtube.com/watch?v=gzLJNNBTQ8w and https://www.youtube.com/watch?v=t7rRF4W-dXg), and led us to think maybe it is better to define separate CRDs for those two intents. 

The `alternatives` section below does explore several possibilities in terms of single CRD (in addition to what you proposed). We could continue the discussion there.

Reply:
Naively, what about a sincle ClusterNetworkPolicy API with something like this:

```
spec:
  canBeOverridden: true | false
```

That's the main difference between the CNP and DNP, right? I don't know if that's actually better... but if the goal is to have fewer `Kind`s then that might do it.

Reply:
the boolean field is one of the alternatives proposed towards the end of this KEP, along with other alternatives which are all single CRD approach.